### DISCUSSION TITLE
.aggrid vs .table - Features, Current & Future Development, Custom Renderers...

### DISCUSSION THREAD

**Discussion by algotr34** [zauberzeug/nicegui#2002] [Q&A] [answered]
Hi
I really like NiceGUI and want to use it for my projects.

I do need some advanced table capabilities (the more customization allowed, the better), and I would like to ask about the differences between: .table or .aggrid, and their features and development status, I wanted to ask:

- Current Status: Based on the current state of NiceGUI,  which offers more options and customization?
- Future Developments: Are you planning to support and develop both? is there a preferred one?
- Does each solution support all of its original options? Can I implement all the options listed in the AG-Grid's documentation in NiceGUI's .aggrid? and same for the Quasar Table: are all the features mentioned in their documentation can be done with .table? If not, do you have a list of which features are supported and which aren't?
- Example: Custom Cell Renderers: I was able to create a Custom Cell Rendering example with .table, but not with .aggrid, although they do support this option: https://www.ag-grid.com/javascript-data-grid/component-cell-renderer/ - but i'm not sure, so I will ask: is the fact that they have this option, means that it should work in nicegui? is there a methodic way to read the ag-grid documentation and implement it into nicegui?

Thanks

**Comment by stdusr** [comment]
Some of these questions were answered here: https://github.com/zauberzeug/nicegui/discussions/1782

**Reply by algotr34** [reply]
Thanks. I did read this thread before asking the question.
The answers that were given by ChatGPT in this thread are about the differences between the separate components (Quasars Table vs AG-Grid), but they are not answering the questions in this current thread, regarding NiceGUI implementation and support of these components.

**Comment by algotr34** [comment]
After doing more research, I think that I will use ag-grid. It has more flexibility and customization.
I am considering using NiceGui for it, but I'm not sure how much of ag-grid functionalities can be used from NiceGui.

For example: I could not find any documentation on how to use the enterprise features. I think it is unsupported, right?
Another example: I could not activate the cell renderer (to put multi-select inside a table cell).
So, I'm just not sure how much of the functionality of ag-grid is/will be supported. Maybe it is too complex to keep it up-to-date?

If 99% of the functionality of ag-grid are supported by NiceGui, I will use NiceGui.
If its close to 50%.... I might have to bypass Nicegui: to use Flask + ag-grid directly (maybe with some framework like vue, just like NiceGui uses). Since i'm more familiar with the python world, It will be painful and annoying, but I need the full functionality...

Anybody knows? does NiceGui have a list of supported/unsupported functionalities of ag-grid?

**Comment by stdusr** [comment]
Yep, enterprise features would be really welcome ... )

**Comment by falkoschindler** [comment, marked as answer]
Hi @algotr34, here are my thoughts on the status of `ui.table` and `ui.aggrid`:

- We don't know either of both frameworks completely. Especially AG Grid is a world of its own and we only scratched the surface with our demos and examples. So there are many unknown unknowns when judging their integration and feature completeness.
- We try to provide a lean interface to access as many of their features as possible. Many NiceGUI users seem to successfully work with tables and AG Grids, so the integration NiceGUI offers turned out pretty powerful.
- Some features are supported with a neat API, others require a bit more custom code. This is common for NiceGUI elements. Translating all, e.g., AG Grid features in a Python API would be a project on its own. But with some generic interfaces like an `options` dictionary or a `call_api_method` we make something like 90%(?) of features accessible with relatively little effort.
- But not _all_ features are technically possible (at the moment, as far as I know). I don't have a specific example in mind, but usually there's always something deep inside of such mighty frameworks which doesn't play nicely with NiceGUI.
- I can't give a list of supported features, because I don't even know all of them and many things are available indirectly via above-mentioned generic interfaces.
- Yes, we're planning to keep supporting both in the foreseeable future.
- Since `ui.table` is based on a Quasar component like many other NiceGUI elements, it feels a bit more "native" and we might prefer to implement a feature there first. But as far as I can tell, AG Grid might be even more powerful already. And AG Grid has a large community. Maybe we find maintainers especially for this element to boost implementation of new features and shortcuts.

To conclude: I think both elements are pretty powerful and can be used in production. If there's a specific feature missing, we're happy to try implementing it, maybe with the help of the community. But we can't guarantee anything. Sometimes there's just some nasty technical limitation. If you need, e.g., AG Grid in its purest form with all features one can think of, an official supported frontend like Javascript, React, Angular, Vue or Solid might be the better choice. But then you'd need to leave Python, so choose wisely! 😉

**Comment by WolfgangFahl** [comment]
If you are interested in ag_grid improvements you might want to look into [nicegui widgets](https://github.com/WolfgangFahl/nicegui_widgets). That project has a [lod_grid module](https://github.com/WolfgangFahl/nicegui_widgets/blob/main/ngwidgets/lod_grid.py) with a [ListOfDictsGrid](https://github.com/WolfgangFahl/nicegui_widgets/blob/22eb57af17fe5d6dbefc7b400a573d356208a579/ngwidgets/lod_grid.py#L12) 
[list of dict support](https://github.com/WolfgangFahl/nicegui_widgets/issues/28) was one of the initial features. Since at BITPlan and our RWTH projects we migrate from justpy to nicegui we have to get our table based code improved as we already did in the past for Flask With Bootstrap5 and for justpy. There is still a lot to do but the discussion and the issues on nicegui widgets are open. Support for different storage backends is a key feature we are looking for. In the past we used [pyLodStorage](https://pypi.org/project/pyLodStorage/) for this purpose.

**Comment by algotr34** [comment]
@falkoschindler Thank you for your detailed answer.
I understand that implementing all the features of ag-grid might be challenging, and that there might be more important things.

For my current project I specifically need the full set of ag-grid features, so I might implement it manually, but I will use NiceGUI for other projects. I really like the simple, short and python-style coding of NiceGUI. It has a lot of potential.
